src/search/search_session.cpp

```diff
@@ -0,0 +1,869 @@
+﻿/*
+===========================================================================
+
+  Copyright (c) 2023 LandSandBoat Dev Teams
+
+  This program is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see http://www.gnu.org/licenses/
+
+===========================================================================
+*/
+
+#include "search_session.h"
+#include "common/md52.h"
+#include "common/socket.h" // for ref<T>
+#include "common/utils.h"  // for unpack/pack bits
+#include "data_loader.h"
+#include <unordered_set>
+
+#include "packets/auction_history.h"
+#include "packets/auction_list.h"
+#include "packets/linkshell_list.h"
+#include "packets/party_list.h"
+#include "packets/search_comment.h"
+#include "packets/search_list.h"
+
+
+// A single IP should only have one request in flight at a time, so we are going to
+// be tracking the IP addresses of incoming requests and if we haven't cleared the
+// record for it - we block until its done
+extern shared_guarded<std::unordered_set<std::string>> gIPAddressesInUse;
+
+// NOTE: We're only using the read-lock for this
+extern shared_guarded<std::unordered_set<std::string>> gIPAddressWhitelist;
+
+search_session::search_session(asio::ip::tcp::socket socket)
+: socket_(std::move(socket))
+{
+    asio::error_code ec = {};
+    socket_.lowest_layer().set_option(asio::socket_base::reuse_address(true));
+    ipAddress = socket_.lowest_layer().remote_endpoint(ec).address().to_string();
+
+    // If someone from the same IP is making a request, block for 10 ms in this thread repeatedly until the request is done.
+    // TODO: do we need a better timeout to drop the request if data never comes from a DIFFERENT socket?
+    // Someone doing this maliciously would just lock themselves out...
+    uint32_t numWaits = 0;
+    while (isIPAddressInUse(ipAddress))
+    {
+        std::this_thread::sleep_for(std::chrono::milliseconds (10));
```

Comment:
Rather than a full thread sleep, can we just use the asio countdown timer/countdown? Then you could keep the ip/numWaits as part of search server or something?

Reply:
Regardless, done. used an asio steady timer